### PR TITLE
Add GLDAS scripts & fix GLDAS job

### LINKED ISSUE #622: gdasgldas task fails with Restart Tile Space Mismatch
**Expected behavior**
gdasgldas task should complete successfully and global-workflow should continue to cycle fv3gdas

**Current behavior**
gdasgldas task is failing on the first cycle in which the task is not skipped (first 00z analysis period when enough data has been produced to trigger task).

**Machines affected**
This error is being expressed on Orion.

**To Reproduce**
I am seeing this bug in a test of global-workflow being conducted on Orion in the following directories:
expid: /work/noaa/da/bhoover/para/bth_test
code: /work/noaa/da/bhoover/global-workflow
ROTDIR: /work/noaa/stmp/bhoover/ROTDIRS/bth_test
RUNDIR: /work/noaa/stmp/bhoover/RUNDIRS/bth_test

This run is initialized on 2020082200, and designed to terminate 2 weeks later on 2020090500.

Experiment setup:
/work/noaa/da/bhoover/global-workflow/ush/rocoto/setup_expt.py --pslot bth_test --configdir  /work/noaa/da/bhoover/global-workflow/parm/config --idate 2020082200 --edate 2020090500 --comrot /work/noaa/stmp/bhoover/ROTDIRS --expdir /work/noaa/da/bhoover/para --resdet 384 --resens 192 --nens 80 --gfs_cyc 1

Workflow setup:
/work/noaa/da/bhoover/global-workflow/ush/rocoto/setup_workflow.py --expdir /work/noaa/da/bhoover/para/bth_test

Initial conditions:
/work/noaa/da/cthomas/ICS/2020082200/

The error is found in the gdasgldas task on 2020082600.

Log file:
/work/noaa/stmp/bhoover/ROTDIRS/bth_test/logs/2020082600/gdasgldas.log

**Context**
This run is being used by a new Orion user and member of the satellite DA group, only to familiarize myself with the process of carrying out an experiment. There have been no code-changes made for this run. I followed directions for cloning and building the global-workflow, and setting up a cycled experiment, from the available wiki:

https://github.com/NOAA-EMC/global-workflow/wiki/

I did not create the initial condition files, they were instead produced for me. The global-workflow repository was cloned on January 25 2022 (d3028b9d8268028226f9c27800fcd6655e9e4bb8)

The task fails with the following error in the log-file:
> 0:  NOAH Restart File Used: noah.rst                                
> 0:            1        1536         768      389408
> 0:  Restart Tile Space Mismatch, Halting..
> 0:  endrun is being called
> 0: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0

The dimension size of 389408 is suspicious, since earlier in the log a different dimension size is referenced, e.g.:
> 0:  MSG: maketiles -- Size of Grid Dimension:      398658  (           0 )

When I search for "389408" in the log-file, it only appears in two places, one is in the Restart Tile Space Mismatch error, and the other is while running exec/gldas_rst, when reporting the results of a FAST_BYTESWAP:

> 216.121 + /work/noaa/da/bhoover/global-workflow/exec/gldas_rst
> 216.121 + 1>& 1 2>& 2
> FAST_BYTESWAP ALGORITHM HAS BEEN USED AND DATA ALIGNMENT IS CORRECT FOR          4  )
>         1536         768           4     9440776
>            2 tmp0_10cmdown GLDAS STC1
>            3 tmp10_40cmdown GLDAS STC2
>            4 tmp40_100cmdown GLDAS STC3
>            5 tmp100_200cmdown GLDAS STC4
>            6 soill0_10cmdown GLDAS SLC1
>            7 soill10_40cmdown GLDAS SLC2
>            8 soill40_100cmdown GLDAS SLC3
>            9 soill100_200cmdown GLDAS SLC4
>           10 soilw0_10cmdown GLDAS SMC1
>           11 soilw10_40cmdown GLDAS SMC2
>           12 soilw40_100cmdown GLDAS SMC3
>           13 soilw100_200cmdown GLDAS SMC4
>           15 landsfc
>           18 vtypesfc
>           71 tmpsfc GLDAS SKNT
>           72 weasdsfc GLDAS SWE
>           79 cnwatsfc GLDAS CMC
>           88 snodsfc GLDAS SNOD
>            1        1536         768      389408
> 216.602 + err=0

I believe that the error is related to the difference in tile-size between these two values.

**Detailed Description**
I have proposed no change or addition to the code for this run.

**Additional Information**
Prior gdasgldas tasks in the run from initialization to 2020082600 were successful, but they were all skipped either because the analysis was for a non-00z period or because the requisite number of cycles had not been completed to allow the task to trigger. There are no successful gdasgldas tasks in this run that I can use to compare to the one that has failed. I have conferred with more experienced EMC users of fv3gdas and the cause of the problem is not obvious.

**Possible Implementation**
I have no implementation plan to offer.

### LINKED ISSUE #1014: Absorb GLDAS scripts
**Description**

As part of the overarching goal of moving global workflow scripts out of component repositories and under global workflow, all of the scripts currently linked in GLDAS should be moved to global-workflow. FYI @HelinWei-NOAA , please let us know of any concerns, questions, or objections. Once we absorb the following scripts you can remove them in your GLDAS repo.

Scripts to be absorbed:
* jobs/
  * JGDAS_ATMOS_GLDAS
* scripts/
  * exgdas_atmos_gldas.sh
* ush/
  * ~~gldas_archive.sh~~ (not bringing into global-workflow, see discussion below https://github.com/NOAA-EMC/global-workflow/issues/1014#issuecomment-1247045987)
  * gldas_forcing.sh
  * gldas_get_data.sh
  * gldas_liscrd.sh
  * ~~gldas_post.sh~~ (not bringing into global-workflow, see discussion below https://github.com/NOAA-EMC/global-workflow/issues/1014#issuecomment-1247045987)
  * gldas_process_data.sh

**Requirements**

All GLDAS jobs, scripts, and ush scripts that are currently linked by global workflow added to the global workflow repo instead and removed from the ignore list.

**Acceptance Criteria (Definition of Done)**

No jobs, scripts, or ush scripts linked from GLDAS.

**Dependencies**

None

### PR DESCRIPTION
**Pull request by KateFriedman-NOAA**
**Description**

This PR:

1. Adds the GLDAS scripts from the GLDAS repo (https://github.com/NOAA-EMC/GLDAS) into global-workflow (as described in issue #1014). The scripts are also
* Updated to conform to recent script cleanup and enhancements (e.g. add preamble).
* The scripts are removed from `.gitignore` and `link_workflow.sh`. 
* References to obsolete platforms (WCOSS_DELL_P3 and WCOSS_C) are removed.
* Updates for CFP were made to use existing CFP variables and remove machine checks.
* Updates were made to address linter/shellcheck warnings.
2. Fixes GLDAS job by updating fix file settings to use updated fix set provided by @HelinWei-NOAA . (Issue #622)

The following scripts are added to global-workflow:

* jobs/
  * JGDAS_ATMOS_GLDAS
* scripts/
  * exgdas_atmos_gldas.sh
* ush/
  * ~~gldas_archive.sh~~ (not bringing into global-workflow, see discussion below https://github.com/NOAA-EMC/global-workflow/issues/1014#issuecomment-1247045987)
  * gldas_forcing.sh
  * gldas_get_data.sh
  * gldas_liscrd.sh
  * ~~gldas_post.sh~~ (not bringing into global-workflow, see discussion below https://github.com/NOAA-EMC/global-workflow/issues/1014#issuecomment-1247045987)
  * gldas_process_data.sh

GLDAS scripts were first added to branch as-is and then updates were made to address the following:

- add preamble
- convert scripts from ksh to bash
- remove backticks and replace with substitutions
- remove unneeded announcement echos
- some general cleanup
- resolve linter/shellcheck warnings

Fixes #622
Resolves #1014

**Type of change**

- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**

- [x] Clone and build tests on Orion
- [x] C192C96L127 cold-started cycled test on Orion (21.5 cycles = 16.5 cycle spinup for GLDAS + 5 cycles after spinup with 2 00z cycles with GLDAS invoked)
  
**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

### CODE REVIEW

**Comment by KateFriedman-NOAA** [comment]
@aerorahul @WalterKolczynski-NOAA This PR is ready for your review. Since initiating it I have:

1. Addressed and resolved all linter/shellcheck warnings.
2. Adjusted how the GLDAS scripts use CFP but incorporating existing CFP variables (`USE_CFP` and `CFP_MP`) to remove machine checks.
3. Fixed GLDAS job (issue #622) by testing updated GLDAS fix files provided by @HelinWei-NOAA and updating GLDAS scripts to use them based on the value of the `FRAC_GRID` variable.

This PR will now resolve issues #622 and #1014.

File: jobs/JGDAS_ATMOS_GLDAS [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+##############################################
+# Obtain unique process id (pid) and make temp directory
+##############################################
+export pid=${pid:-$$}
+export outid=${outid:-"LL${job}"}
+
+export DATA=${DATA:-${DATAROOT}/${jobid:?}}
+mkdir -p "${DATA}"
+cd "${DATA}" || exit
```
Comment (WalterKolczynski-NOAA) [inline jobs/JGDAS_ATMOS_GLDAS]: This should be a non-zero exit since it is an error.

File: jobs/JGDAS_ATMOS_GLDAS [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+# Final processing
+##############################################
+if [[ -e "${pgmout}" ]] ; then
+  cat "${pgmout}"
+fi
+
+##########################################
+# Remove the Temporary working directory
+##########################################
+cd "${DATAROOT}" || exit
```
Comment (WalterKolczynski-NOAA) [inline jobs/JGDAS_ATMOS_GLDAS]: Theoretically so should this one, but I don't know if we want to job to "fail" it it just can't delete `${DATA}`.

File: parm/config/config.fcst [hunk 3]
```diff
@@ -252,9 +252,8 @@ export DO_OCN_SPPT=${DO_OCN_SPPT:-"NO"}
 export DO_OCN_PERT_EPBL=${DO_OCN_PERT_EPBL:-"NO"}
 
 #coupling settings
-export FRAC_GRID=".true."
 export cplmode="nems_frac"
-if [ $FRAC_GRID = ".false." ]; then
+if [[ "${FRAC_GRID:-".true."}" = ".false." ]]; then
```
Comment (WalterKolczynski-NOAA) [inline parm/config/config.fcst:256]: Does this work as intended? Seems like there might be a quotation mark issue on the left side.

File: scripts/exgdas_atmos_gldas.sh [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+EOF
+cp fort.141 fort.41
+
+
+# 3b) Use gdas2gldas to generate nemsio file 
+
+export OMP_NUM_THREADS=1
+export pgm=gdas2gldas
+# shellcheck disable=SC1091
+. prep_step
```
Comment (WalterKolczynski-NOAA) [inline scripts/exgdas_atmos_gldas.sh:198]: We should probably look at whether this actually needs to be sourced, or just run normally. And does it actually have to be called before each step? Can be later though.

File: ush/gldas_forcing.sh [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+export WGRIB=${WGRIB:?}
+export COPYGB=${COPYGB:?}
+export ERRSCRIPT=${ERRSCRIPT:-"eval [[ ${err} = 0 ]]"}
+
+#-------------------------------
+#--- extract variables of each timestep and create forcing files
+sdate=${bdate}
+edate=$(sh "${FINDDATE:?}" "${edate}" d-1)
+while [[ "${sdate}" -lt "${edate}" ]] ; do
+#-------------------------------
```
Comment (WalterKolczynski-NOAA) [inline ush/gldas_forcing.sh]: Indent block

File: ush/gldas_get_data.sh [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+#-------------------------------
+while [[ "${cdate}" -lt "${edate}" ]]; do
+#-------------------------------
+  ymd=$(echo "${cdate}" |cut -c 1-8)
+  cyc=$(echo "${cdate}" |cut -c 9-10)
+  [[ ! -d ${fpath}/gdas.${ymd} ]] && mkdir -p "${fpath}/gdas.${ymd}"
+  [[ ! -d ${gpath}/gdas.${ymd} ]] && mkdir -p "${gpath}/gdas.${ymd}"
+
+f=1
```
Comment (WalterKolczynski-NOAA) [inline ush/gldas_get_data.sh]: This is still part of the while block but lost its indentation.

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

File: parm/config/config.fcst (hunk 3 above)
Comment (KateFriedman-NOAA) [reply parm/config/config.fcst:256]: I reran the gdasfcst job and it appears to work. Here are the lines from the log where line 256 in config.fcst is invoked:

```
++ config.fcst[255]: export cplmode=nems_frac
++ config.fcst[255]: cplmode=nems_frac
++ config.fcst[256]: [[ .true. = \.\f\a\l\s\e\. ]]
++ config.fcst[259]: export psm_bc=1
```

File: scripts/exgdas_atmos_gldas.sh (hunk 4 above)
Comment (KateFriedman-NOAA) [reply scripts/exgdas_atmos_gldas.sh:198]: Will check in later work.

File: ush/gldas_forcing.sh (hunk 5 above)
Comment (KateFriedman-NOAA) [reply ush/gldas_forcing.sh]: Indented.

File: ush/gldas_get_data.sh (hunk 6 above)
Comment (KateFriedman-NOAA) [reply ush/gldas_get_data.sh]: Indented.

**Review by WalterKolczynski-NOAA — APPROVED** [review]

File: scripts/exgdas_atmos_gldas.sh [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+# Script description:  Runs the global land analysis                    
+#
+################################################################################
+
+source "${HOMEgfs:?}/ush/preamble.sh"
+
+#################################
+# Set up UTILITIES
+#################################
+export FINDDATE=${FINDDATE:-/apps/ops/prod/nco/core/prod_util.v2.0.13/ush/finddate.sh}
```
Comment (aerorahul) [inline scripts/exgdas_atmos_gldas.sh:16]: Why are wcoss2 paths hard-wired here and below?
I guess we can handle that in another PR.
Would you add a comment 
```
#TODO - remove hard-wired WCOSS2 default paths
```

**Review by aerorahul — APPROVED** [review]
Mostly looks good.
This PR basically brings in scripts from the GLDAS repo into global-workflow, and applies the linter recommended standards.

File: scripts/exgdas_atmos_gldas.sh (hunk 7 above)
Comment (KateFriedman-NOAA) [reply scripts/exgdas_atmos_gldas.sh:16]: Good catch. Those defaults should use module-provided paths to those utilities. I will correct in a subsequent PR. Thanks!

File: jobs/JGDAS_ATMOS_GLDAS (hunk 1 above)
Comment (KateFriedman-NOAA) [reply jobs/JGDAS_ATMOS_GLDAS]: Added "1" with the exit:

```
cd "${DATA}" || exit 1
```

File: jobs/JGDAS_ATMOS_GLDAS (hunk 2 above)
Comment (KateFriedman-NOAA) [reply jobs/JGDAS_ATMOS_GLDAS]: Added "1" with the exit:

```
cd "${DATA}" || exit 1
```

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Pull request merged**